mediatek: add support for Mercusys MR80X v3

Schneider-Azima opened the pull request:
This commit adds support for Mercusys MR80X(EU) v3 router.

Device specification:
 - SoC: Mediatek MT7981b, Cortex-A53, 64-bit
 - RAM: 512MB
 - Flash: SPI NAND GigaDevice GD5F1GQ5UEYIGY (128 MB)
 - Ethernet: 4x 100/1000 Mbps LAN1,LAN2,LAN3 & WAN
 - Wireless: 2.4GHz (802.11 b/g/n/ax)
 - Wireless: 5GHz (802.11 a/n/ac/ax)
 - LEDs: 1 orange and 1 green status LEDs, 4 green gpio-controlled LEDs on ethernet ports
 - Buttons: 1 (Reset)
 - Bootloader: Main U-Boot - U-Boot 2022.01-rc4. Additionally, both UBI slots contain "seconduboot" (also U-Boot 2022.01-rc4)

Installation (UART):
 - Place OpenWrt sysupgrade image on tftp server with IP 192.168.1.2
 - Attach UART, switch on the router and interrupt the boot process by pressing 'Ctrl-C'.
 - Upload the sysupgrade.bin via tftp. setenv serverip 192.168.1.2; setenv ipaddr 192.168.1.1; tftp 0x40001000 sysupgrade.bin
 - Erase "ubi0" part, and write the image. nand erase.spread 0x300000 0x3200000; nand write.spread 0x40001000 0x300000 0x3200000
 - Set the uboot environment for startup. setenv tp_boot_idx 0; setenv bootcmd bootm 0x46000000; saveenv
 - Reboot the router by reconnecting the power.

Recovery:
 - Press Reset button and power on the router.
 - Navigate to U-Boot recovery web server (http://192.168.1.1/) and upload the OEM firmware.

Stock layout:
0x000000000000-0x000000200000 : "boot"
0x000000200000-0x000000300000 : "u-boot-env"
0x000000300000-0x000003500000 : "ubi0"
0x000003500000-0x000006700000 : "ubi1"
0x000006700000-0x000006f00000 : "userconfig"
0x000006f00000-0x000007300000 : "tp_data"

ubi0/ubi1 format:
U-Boot at boot checks that all volumes are in place: +-------------------------------+
| Volume Name: uboot   Vol ID: 0|
| Volume Name: kernel  Vol ID: 1|
| Volume Name: rootfs  Vol ID: 2|
+-------------------------------+

MAC addresses:
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| label   | 94:0C:xx:xx:xx:12 | label     |
| WLAN 2g | 94:0C:xx:xx:xx:11 | label-1   |
| WLAN 5g | 94:0C:xx:xx:xx:10 | label-2   |
+---------+-------------------+-----------+
label MAC address was found in UBI partition "tp_data", file
"default-mac". 

Signed-off-by: Schneider Azima <Schneider-Azima12@protonmail.com>

blogic commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,351 @@
+	model = "MediaTek MT7981 RFB";
+	compatible = "mercusys,mr80x-v3";
+	chosen {
+		//bootargs = "console=ttyS0,115200n1 loglevel=8  \
```
drop the comments

blogic commented on package/boot/uboot-mediatek/patches/445_add_mercusys_mr80xv3_dts_and_target.patch:
```diff
@@ -0,0 +1,326 @@
+--- /dev/null
++++ b/configs/mt7981_mercusys_mr80xv3_defconfig
```
this is not needed i guess

blogic commented on package/boot/uboot-mediatek/patches/445_add_mercusys_mr80xv3_dts_and_target.patch:
```diff
@@ -0,0 +1,326 @@
+--- /dev/null
++++ b/configs/mt7981_mercusys_mr80xv3_defconfig
+@@ -0,0 +1,187 @@
```
the uboot part needs to go into a separate patch

blogic commented on package/kernel/mt76/Makefile:
```diff
@@ -617,6 +617,7 @@ define KernelPackage/mt7981-firmware/install
 		$(PKG_BUILD_DIR)/firmware/mt7981_wa.bin \
 		$(PKG_BUILD_DIR)/firmware/mt7981_wm.bin \
 		$(PKG_BUILD_DIR)/firmware/mt7981_rom_patch.bin \
```
mt76 changes need to go into a separate patch

Schneider-Azima committed:
mediatek: add support for Mercusys MR80X v3

This commit adds support for Mercusys MR80X(EU) v3 router.

Device specification:
 - SoC: Mediatek MT7981b, Cortex-A53, 64-bit
 - RAM: 512MB
 - Flash: SPI NAND GigaDevice GD5F1GQ5UEYIGY (128 MB)
 - Ethernet: 4x 100/1000 Mbps LAN1,LAN2,LAN3 & WAN
 - Wireless: 2.4GHz (802.11 b/g/n/ax)
 - Wireless: 5GHz (802.11 a/n/ac/ax)
 - LEDs: 1 orange and 1 green status LEDs, 4 green gpio-controlled LEDs
   on ethernet ports
 - Buttons: 1 (Reset)
 - Bootloader: Main U-Boot - U-Boot 2022.01-rc4. Additionally, both UBI
   slots contain "seconduboot" (also U-Boot 2022.01-rc4)

Installation (UART):
 - Place OpenWrt sysupgrade image on tftp server with IP 192.168.1.2
 - Attach UART, switch on the router and interrupt the boot process by
   pressing 'Ctrl-C'.
 - Upload the sysupgrade.bin via tftp.
   setenv serverip 192.168.1.2; setenv ipaddr 192.168.1.1; tftp
    0x40001000 sysupgrade.bin
 - Erase "ubi0" part, and write the image.
   nand erase.spread 0x300000 0x3200000; nand write.spread 0x40001000
    0x300000 0x3200000
 - Set the uboot environment for startup.
   setenv tp_boot_idx 0; setenv bootcmd bootm 0x46000000; saveenv
 - Reboot the router by reconnecting the power.

Recovery:
 - Press Reset button and power on the router.
 - Navigate to U-Boot recovery web server (http://192.168.1.1/) and
   upload the OEM firmware.

Stock layout:
0x000000000000-0x000000200000 : "boot"
0x000000200000-0x000000300000 : "u-boot-env"
0x000000300000-0x000003500000 : "ubi0"
0x000003500000-0x000006700000 : "ubi1"
0x000006700000-0x000006f00000 : "userconfig"
0x000006f00000-0x000007300000 : "tp_data"

ubi0/ubi1 format:
U-Boot at boot checks that all volumes are in place:
+-------------------------------+
| Volume Name: uboot   Vol ID: 0|
| Volume Name: kernel  Vol ID: 1|
| Volume Name: rootfs  Vol ID: 2|
+-------------------------------+

MAC addresses:
+---------+-------------------+-----------+
|         | MAC               | Algorithm |
+---------+-------------------+-----------+
| label   | 94:0C:xx:xx:xx:12 | label     |
| WLAN 2g | 94:0C:xx:xx:xx:11 | label-1   |
| WLAN 5g | 94:0C:xx:xx:xx:10 | label-2   |
+---------+-------------------+-----------+
label MAC address was found in UBI partition "tp_data", file
"default-mac".

Signed-off-by: Schneider Azima <Schneider-Azima12@protonmail.com>

Schneider-Azima committed:
uboot-mediatek: add MERCUSYS MR80X v3 support

Add U-Boot replacement loader for the MERCUSYS MR80X. This is required
to increase available flash space in OpenWrt.

Signed-off-by: Schneider Azima <Schneider-Azima12@protonmail.com>

Schneider-Azima committed:
mt76: add missing eeprom file for mt7981

Mercusys MR80x need this eeprom to enable WiFi.

Signed-off-by: Schneider Azima <Schneider-Azima12@protonmail.com>

blogic commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,350 @@
+/dts-v1/;
```
The SPDX license header is missing, please add it

hauke commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,352 @@
+
+#include "mt7981.dtsi"
+/ {
+	model = "MediaTek MT7981 RFB";
```
Please add the name of the device here.

hauke commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,352 @@
+#include "mt7981.dtsi"
+/ {
+	model = "MediaTek MT7981 RFB";
+	compatible = "mercusys,mr80x-v3";
```
```suggestion
	compatible = "mercusys,mr80x-v3", "mediatek,mt7981";
```

hauke commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,352 @@
+		compatible = "gpio-leds";
+
+		orange {
+			label = "orange";
```
Please use the function definition, for example:
```
			function = LED_FUNCTION_WLAN_2GHZ;
			color = <LED_COLOR_ID_WHITE>;
```

hauke commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,352 @@
+		};
+	};
+
+	sound_wm8960 {
+		compatible = "mediatek,mt79xx-wm8960-machine";
+		mediatek,platform = <&afe>;
+		audio-routing = "Headphone", "HP_L",
+				"Headphone", "HP_R",
+				"LINPUT1", "AMIC",
+				"RINPUT1", "AMIC";
+		mediatek,audio-codec = <&wm8960>;
+		status = "disabled";
+	};
+
+	sound_si3218x {
+		compatible = "mediatek,mt79xx-si3218x-machine";
+		mediatek,platform = <&afe>;
+		mediatek,ext-codec = <&proslic_spi>;
+		status = "disabled";
+	};
```
Does the board support sound output?

hauke commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,352 @@
+	pinctrl-0 = <&i2c_pins>;
+	status = "disabled";
+
+	wm8960: wm8960@1a {
+		compatible = "wlf,wm8960";
+		reg = <0x1a>;
+	};
```
This is an audio codec. Is this really on the board?

hauke commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,352 @@
+		nick-index = <4>;
+		phy-handle = <&phy0>;
+		phy-link {
+            speed = <1000>;
+            full-duplex;
```
indention is wrong

hauke commented on target/linux/mediatek/dts/mt7981b-mercusys-mr80x-v3.dts:
```diff
@@ -0,0 +1,352 @@
+			partner_gmac = <1>;
+		};
+		phy0: ethernet-phy@0 {
+			compatible = "ethernet-phy-id03a2.9461";
```
I prefer when you just provide the generic c22 or c45 name:
```suggestion
			compatible = "ethernet-phy-ieee802.3-c22";
```

hauke replied:
The `phy0: ethernet-phy@0 ` entry is not needed, it is already in the `target/linux/mediatek/files-6.6/arch/arm64/boot/dts/mediatek/mt7981.dtsi` file.

fede843 commented:
Hi, what is the status on this one? any chance to be included in 24.10?

developer-iname commented:
Hi, what model is this?

I have the V1.2 EU model and i'm having alot of trouble getting any output from the uart serial interface.

I am new to using UART to flashing routers and stuff, I found the ground pin with my multi meter and then tried to check for the board tx pin with my RX interface and never got a good output out of the other 3 pins I used, even tried rx2322 and rxd

Schneider-Azima commented:
> Hi, what is the status on this one? any chance to be included in 24.10?

I've changed the PR according to all above requirements. I'm waiting for 6 approving reviewers. Is there a way to remind other reviewers to check this PR?

Schneider-Azima commented:
There seems to be various models of MR80X according to openwrt forum. And your V1.2EU probably isn't MR80X V3 EU. Maybe you can check the CPU model after connecting the serial port. It may be IPQ or MTK, and MR80X V3 EU uses MTK.

As for your serial problem, I'm not exactly sure your connecting method. For your information, there usually be four pins as TX/RX/GND/VCC.

通过 [Proton Mail](https://proton.me/mail/home) 安全邮箱发送。

2025年2月4日 星期二 上午 3:58，Ross Sadler ***@***.***> 来信：

> Hi, what model is this?
>
> I have the V1.2 EU model and i'm having alot of trouble getting any output from the uart serial interface.
>
> I am new to using UART to flashing routers and stuff, I found the ground pin with my multi meter and then tried to check for the board tx pin with my RX interface and never got a good output out of the other 3 pins I used, even tried rx2322 and rxd
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/openwrt/openwrt/pull/17169#issuecomment-2631944104), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/BLBHI3KEB24YEOC6KIW3XRL2N7C6NAVCNFSM6AAAAABTB2CGUGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMZRHE2DIMJQGQ).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

Schneider-Azima commented:
There seems to be various models of MR80X according to openwrt forum. And your V1.2EU probably isn't MR80X V3 EU. Maybe you can check the CPU model after connecting the serial port. It may be IPQ or MTK, and MR80X V3 EU uses MTK.
As for your serial problem, I'm not exactly sure your connecting method. For your information, there usually be four pins as TX/RX/GND/VCC.

> Hi, what model is this?
>
> I have the V1.2 EU model and i'm having alot of trouble getting any output from the uart serial interface.
>
> I am new to using UART to flashing routers and stuff, I found the ground pin with my multi meter and then tried to check for the board tx pin with my RX interface and never got a good output out of the other 3 pins I used, even tried rx2322 and rxdMessage ID: ***@***.***>

Schneider-Azima commented:
@blogic @hauke Any chance to be merged in 24.10? Do I need to format a new PR to clean the labels and sync with main branch?

Schneider-Azima commented:
I create a new pull request #18181 . And I made two changes.
1. Use stock eeprom instead of compiled prooduct of mt76, as what have done in mr90x #13575  and ax80v1 #17753 .
2. Add web upgrade support.

I kindly invite you to review the new PR and suggest changes. @blogic @hauke @fede843

hauke commented:
Support for this device was merged in 7921e48d43571861f15aa56717798709ff7507b6 using #18181
